File: tests/multilingual_test.py
# Python 测试文件 - 包含各种Python特性

import asyncio
from typing import List, Dict, Optional, Union
from dataclasses import dataclass
from abc import ABC, abstractmethod
import json

# 数据类装饰器
@dataclass
class Person:
    name: str
    age: int
    email: str = ""

# 普通类定义
class Animal:
    """动物基类"""
    
    def __init__(self, name: str, species: str):
        self.name = name
        self.species = species
    
    def speak(self) -> str:
        return f"{self.name} makes a sound"

# 抽象类
class AbstractVehicle(ABC):
    @abstractmethod
    def start_engine(self) -> str:
        pass
    
    @abstractmethod
    def stop_engine(self) -> str:
        pass

# 继承类
class Car(AbstractVehicle):
    def __init__(self, brand: str, model: str):
        self.brand = brand
        self.model = model
        self._engine_running = False
    
    def start_engine(self) -> str:
        self._engine_running = True
        return f"{self.brand} {self.model} engine started"
    
    def stop_engine(self) -> str:
        self._engine_running = False
        return f"{self.brand} {self.model} engine stopped"

# 枚举类（Python 3.4+）
from enum import Enum, auto

class Color(Enum):
    RED = auto()
    GREEN = auto()
    BLUE = auto()

class Status(Enum):
    PENDING = "pending"
    ACTIVE = "active"
    COMPLETED = "completed"

# 函数定义
def greet(name: str) -> str:
    """简单的问候函数"""
    return f"Hello, {name}!"

def calculate_area(length: float, width: float) -> float:
    """计算矩形面积"""
    return length * width

def process_data(data: List[Dict[str, Union[str, int]]]) -> Dict[str, int]:
    """处理数据列表"""
    result = {}
    for item in data:
        key = item.get("category", "unknown")
        value = item.get("value", 0)
        result[key] = result.get(key, 0) + value
    return result

# 异步函数
async def fetch_data(url: str) -> Dict:
    """模拟异步数据获取"""
    await asyncio.sleep(1)  # 模拟网络延迟
    return {"url": url, "data": "sample data"}

async def main_async():
    """主异步函数"""
    result = await fetch_data("https://api.example.com/data")
    return result

# 生成器函数
def fibonacci_generator(n: int):
    """斐波那契数列生成器"""
    a, b = 0, 1
    for _ in range(n):
        yield a
        a, b = b, a + b

# Lambda函数
square = lambda x: x ** 2
cube = lambda x: x ** 3

# 高阶函数
def create_multiplier(factor: int):
    """创建乘法器函数"""
    def multiplier(x: int) -> int:
        return x * factor
    return multiplier

times_three = create_multiplier(3)

# 上下文管理器
class FileManager:
    """文件上下文管理器"""
    
    def __init__(self, filename: str, mode: str):
        self.filename = filename
        self.mode = mode
        self.file = None
    
    def __enter__(self):
        self.file = open(self.filename, self.mode)
        return self.file
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        if self.file:
            self.file.close()

# 属性装饰器
class Circle:
    def __init__(self, radius: float):
        self._radius = radius
    
    @property
    def radius(self) -> float:
        return self._radius
    
    @radius.setter
    def radius(self, value: float):
        if value > 0:
            self._radius = value
    
    @property
    def area(self) -> float:
        return 3.14159 * self._radius ** 2

# 静态方法和类方法
class MathUtils:
    PI = 3.14159
    
    @staticmethod
    def add(x: float, y: float) -> float:
        return x + y
    
    @classmethod
    def get_pi(cls) -> float:
        return cls.PI

# 异常处理
def safe_divide(a: float, b: float) -> float:
    """安全的除法运算"""
    try:
        return a / b
    except ZeroDivisionError:
        print("Error: Division by zero")
        return 0.0
    except TypeError:
        print("Error: Invalid type")
        return 0.0

# 列表推导式
def process_numbers(numbers: List[int]) -> List[int]:
    """使用列表推导式处理数字"""
    return [x ** 2 for x in numbers if x > 0]

# 字典推导式
def create_square_dict(n: int) -> Dict[int, int]:
    """创建平方数字典"""
    return {i: i ** 2 for i in range(1, n + 1)}

# 主函数
def main():
    print("Python Multilingual Test")
    
    # 基本函数测试
    print(greet("World"))
    print(f"Area: {calculate_area(5.0, 3.0)}")
    
    # 类测试
    car = Car("Toyota", "Camry")
    print(car.start_engine())
    print(car.stop_engine())
    
    # 枚举测试
    print(f"Color: {Color.RED.name}")
    print(f"Status: {Status.ACTIVE.value}")
    
    # 生成器测试
    fib_sequence = list(fibonacci_generator(10))
    print(f"Fibonacci: {fib_sequence}")
    
    # Lambda测试
    print(f"Square of 5: {square(5)}")
    print(f"Cube of 3: {cube(3)}")
    
    # 高阶函数测试
    print(f"Times three of 4: {times_three(4)}")
    
    # 数据处理测试
    sample_data = [
        {"category": "A", "value": 10},
        {"category": "B", "value": 20},
        {"category": "A", "value": 15}
    ]
    result = process_data(sample_data)
    print(f"Processed data: {result}")
    
    # 列表推导式测试
    numbers = [1, 2, 3, 4, 5, -1, -2]
    squared = process_numbers(numbers)
    print(f"Squared positive numbers: {squared}")
    
    # 字典推导式测试
    square_dict = create_square_dict(5)
    print(f"Square dictionary: {square_dict}")
    
    # 属性测试
    circle = Circle(5.0)
    print(f"Circle radius: {circle.radius}, area: {circle.area}")
    
    # 静态方法测试
    print(f"MathUtils.add(10, 20): {MathUtils.add(10, 20)}")
    print(f"MathUtils.get_pi(): {MathUtils.get_pi()}")
    
    # 异常处理测试
    result = safe_divide(10, 0)
    print(f"Safe divide result: {result}")

# 数据类实例
person1 = Person("Alice", 30, "alice@example.com")

# 运行主函数
if __name__ == "__main__":
    main()

File: tests/multilingual_test.ts
// TypeScript 测试文件 - 包含各种TypeScript特性

// 接口定义
interface Person {
    name: string;
    age: number;
    email?: string;
}

interface Vehicle {
    brand: string;
    model: string;
    start(): string;
    stop(): string;
}

// 类型别名
type Status = 'pending' | 'active' | 'completed';
type ID = string | number;
type Callback<T> = (data: T) => void;

// 枚举
enum Color {
    Red = 'RED',
    Green = 'GREEN',
    Blue = 'BLUE'
}

enum HttpStatus {
    OK = 200,
    NotFound = 404,
    InternalServerError = 500
}

// 类定义
class Animal {
    protected name: string;
    
    constructor(name: string) {
        this.name = name;
    }
    
    speak(): string {
        return `${this.name} makes a sound`;
    }
}

class Dog extends Animal {
    private breed: string;
    
    constructor(name: string, breed: string) {
        super(name);
        this.breed = breed;
    }
    
    bark(): string {
        return `${this.name} the ${this.breed} barks!`;
    }
    
    // 静态方法
    static createPuppy(name: string): Dog {
        return new Dog(name, 'Puppy');
    }
}

// 抽象类
abstract class AbstractShape {
    abstract getArea(): number;
    abstract getPerimeter(): number;
    
    describe(): string {
        return `Shape with area ${this.getArea()}`;
    }
}

class Rectangle extends AbstractShape {
    constructor(private width: number, private height: number) {
        super();
    }
    
    getArea(): number {
        return this.width * this.height;
    }
    
    getPerimeter(): number {
        return 2 * (this.width + this.height);
    }
}

// 泛型类
class Stack<T> {
    private items: T[] = [];
    
    push(item: T): void {
        this.items.push(item);
    }
    
    pop(): T | undefined {
        return this.items.pop();
    }
    
    peek(): T | undefined {
        return this.items[this.items.length - 1];
    }
    
    isEmpty(): boolean {
        return this.items.length === 0;
    }
}

// 泛型函数
function identity<T>(arg: T): T {
    return arg;
}

function mapArray<T, U>(array: T[], fn: (item: T) => U): U[] {
    return array.map(fn);
}

// 箭头函数
const add = (a: number, b: number): number => a + b;
const multiply = (x: number, y: number): number => x * y;

// 函数重载（通过联合类型实现）
function processValue(value: string): string;
function processValue(value: number): number;
function processValue(value: string | number): string | number {
    if (typeof value === 'string') {
        return value.toUpperCase();
    } else {
        return value * 2;
    }
}

// 可选参数和默认参数
function greet(name: string, greeting?: string): string {
    const actualGreeting = greeting || 'Hello';
    return `${actualGreeting}, ${name}!`;
}

function createUser(name: string, age: number = 18): Person {
    return { name, age };
}

// 装饰器函数（TypeScript装饰器）
function log(target: any, propertyKey: string, descriptor: PropertyDescriptor) {
    const originalMethod = descriptor.value;
    descriptor.value = function (...args: any[]) {
        console.log(`Calling ${propertyKey} with args:`, args);
        const result = originalMethod.apply(this, args);
        console.log(`Method ${propertyKey} returned:`, result);
        return result;
    };
}

function measure(target: any, propertyKey: string, descriptor: PropertyDescriptor) {
    const originalMethod = descriptor.value;
    descriptor.value = function (...args: any[]) {
        const start = performance.now();
        const result = originalMethod.apply(this, args);
        const end = performance.now();
        console.log(`Method ${propertyKey} took ${end - start} milliseconds`);
        return result;
    };
}

// 使用装饰器的类
class Calculator {
    @log
    @measure
    add(a: number, b: number): number {
        return a + b;
    }
    
    multiply(a: number, b: number): number {
        return a * b;
    }
}

// 异步函数
async function fetchData(url: string): Promise<any> {
    // 模拟异步操作
    return new Promise((resolve) => {
        setTimeout(() => {
            resolve({ url, data: 'sample data' });
        }, 1000);
    });
}

async function processAsync(): Promise<void> {
    try {
        const result = await fetchData('https://api.example.com/data');
        console.log('Async result:', result);
    } catch (error) {
        console.error('Async error:', error);
    }
}

// 高级类型
type User = {
    id: number;
    name: string;
    email: string;
    roles: string[];
};

type AdminUser = User & {
    permissions: string[];
    isActive: boolean;
};

type Nullable<T> = T | null;
type Optional<T> = T | undefined;

// 条件类型
type IsString<T> = T extends string ? true : false;
type ExtractArrayType<T> = T extends (infer U)[] ? U : never;

// 映射类型
type Readonly<T> = {
    readonly [P in keyof T]: T[P];
};

type Partial<T> = {
    [P in keyof T]?: T[P];
};

// 命名空间
namespace Utils {
    export function formatDate(date: Date): string {
        return date.toISOString().split('T')[0];
    }
    
    export class StringUtils {
        static capitalize(str: string): string {
            return str.charAt(0).toUpperCase() + str.slice(1);
        }
        
        static reverse(str: string): string {
            return str.split('').reverse().join('');
        }
    }
}

// 模块导出
export interface ILogger {
    log(message: string): void;
    error(message: string): void;
}

export class ConsoleLogger implements ILogger {
    log(message: string): void {
        console.log(`[LOG] ${message}`);
    }
    
    error(message: string): void {
        console.error(`[ERROR] ${message}`);
    }
}

// 混合类型
interface Timestamped {
    createdAt: Date;
    updatedAt: Date;
}

interface Identifiable {
    id: string;
}

type TimestampedEntity = Timestamped & Identifiable;

// 实用工具类型
type KeysOfType<T, U> = {
    [K in keyof T]: T[K] extends U ? K : never;
}[keyof T];

type PickByType<T, U> = Pick<T, KeysOfType<T, U>>;

// 主函数
function main(): void {
    console.log("TypeScript Multilingual Test");
    
    // 基本函数测试
    console.log(greet("World"));
    console.log(add(5, 3));
    console.log(multiply(4, 7));
    
    // 类测试
    const dog = new Dog("Buddy", "Golden Retriever");
    console.log(dog.bark());
    console.log(dog.speak());
    
    const puppy = Dog.createPuppy("Max");
    console.log(puppy.bark());
    
    // 泛型测试
    const numberStack = new Stack<number>();
    numberStack.push(1);
    numberStack.push(2);
    console.log(`Stack peek: ${numberStack.peek()}`);
    
    const stringStack = new Stack<string>();
    stringStack.push("hello");
    stringStack.push("world");
    console.log(`Stack peek: ${stringStack.peek()}`);
    
    // 枚举测试
    console.log(`Color: ${Color.Red}`);
    console.log(`HTTP Status: ${HttpStatus.OK}`);
    
    // 函数重载测试
    console.log(`Process string: ${processValue("hello")}`);
    console.log(`Process number: ${processValue(42)}`);
    
    // 可选参数和默认参数测试
    console.log(greet("Alice"));
    console.log(greet("Bob", "Hi"));
    
    const user = createUser("Charlie", 25);
    console.log(`User: ${JSON.stringify(user)}`);
    
    // 装饰器测试
    const calc = new Calculator();
    console.log(`Add result: ${calc.add(10, 20)}`);
    console.log(`Multiply result: ${calc.multiply(5, 6)}`);
    
    // 命名空间测试
    console.log(`Formatted date: ${Utils.formatDate(new Date())}`);
    console.log(`Capitalized: ${Utils.StringUtils.capitalize("hello")}`);
    console.log(`Reversed: ${Utils.StringUtils.reverse("world")}`);
    
    // 高级类型测试
    const adminUser: AdminUser = {
        id: 1,
        name: "Admin",
        email: "admin@example.com",
        roles: ["admin", "user"],
        permissions: ["read", "write", "delete"],
        isActive: true
    };
    
    console.log(`Admin user: ${JSON.stringify(adminUser)}`);
    
    // 异步测试（注释掉以避免运行时依赖）
    // processAsync().then(() => console.log("Async processing completed"));
}

// 导出函数和类
export { main, Person, Vehicle, Calculator, Stack, Color, Status };
export default main;